### ISSUE TITLE
Feature Request: Allow Resolvers section with non existing dns servers

### ISSUE THREAD

**Issue by f1-outsourcing** [haproxy/haproxy#807]
Haproxy 2.2 is designed not to start if already one nameserver cannot be reached due to routing issues.

[ALERT] 220/115919 (3924259) : config : resolvers ‘dnssvrs1’: can’t connect socket for nameserver ‘aaa.aa.aaa.aaa’.
[ALERT] 220/115919 (3924259) : Fatal errors found in configuration.

see discussed thread here
https://discourse.haproxy.org/t/resolvers-section-with-non-existing-dns-servers-possible/5549/

**Comment by lukastribus** [comment]
What happens currently is that if the routing stack returns `No route to host` for example, leading to a `EHOSTUNREACH` response to the connect(), haproxy will refuse to start.

(reproducable with `route add -host <IP-ADDRESS> reject `)

While it's good to know, refusing to start could be too extreme. Downgrading from a critical alert to a warning would be a possibility.

**Comment by bedis** [comment]
what should we do when a resolvers section has no DNS server available? Still warning or do report an alert?
This might require as well a new status on the server such as "no nameserver available for resolution"

**Comment by roidelapluie** [comment]
I do not like this change, it sounds like an bug to me. The current behavior seems correct.

**Comment by roidelapluie** [comment]
I do not think that refusing to start is extreme, as it is the users who writes the configuration file. Having wrong DNS servers here is a bad configuration.

**Comment by lukastribus** [comment]
Currently we have no idea whether a nameserver is available or not. It's not like we health-check nameservers here. The only thing we do is check for `connect()` errors, which only cover some specific errors from the routing stack, that's it.

Let's not pretend that the current code gives us a real indication about nameserver availability.

**Comment by roidelapluie** [comment]
I do not pretend that.

**Comment by bedis** [comment]
As far as I remember, in the first version of the resolver, we did not do the connect at configuration parsing time and users complained that HAProxy could start despite the DNS server was not available :)

 I need to check as well if HAProxy can re-connect() to a server that was unreachable at configuration parsing time.

Based on this, we might add a new option on the nameserver line or in the resolvers section which could say "my DNS servers might be DOWN at configuration parsing time, so please just WARN and don't exit". The idea is that we don't break current behavior and we give the ability for the few people who needs it to be able to workaround it.

**Comment by f1-outsourcing** [comment]
@bedis 
The majority of users are on the level of wordpress admins with bad logics. You will have lots of people with simplistic approach like @roidelapluie. You should stay close to the core of the idea of high availability. 

1.
If you have to choose between running with 4 nameservers instead of 5, or not run at all. I would choose running (see discussion).
Many do not even have own nameservers, so they do not control availability of them, that is even why you specify multiple dns servers. 

2.
Routing is also not an argument, because when you can easily start a container on a different network, lets say a testing network, that does not allow outside access. Then you have to remove your external nameservers from the config just to do testing. Testing config != production config is not really nice.

3. 
Stick to what the 'big' boys are doing. If apache httpd is behaving like this, you should too. They probably have a more experienced and knowledgeable team that already thought this thru. Furthermore you behave according to standards and that is whay sysadmins expect.

Just ask yourself the question every time you make a change. If I have 20 different services and 10.000 requests per second. Would my change allow them to service all requests correctly, service a majority of these requests correctly or should I prevent all these requests from being serviced.

I hope to see this change soon.
